### PR TITLE
[#8370] Add support for Elliptic Curve in twisted.conch.ssh.

### PR DESCRIPTION
**Pull request by glyph**
See: https://twistedmatrix.com/trac/ticket/8370

### CODE REVIEW

**Comment by glyph** [comment]
Please remove the commits that add `trial_out.txt` from the final version of this PR - 16k lines of test output is not something we want sitting around in history forever :)

File: twisted/conch/ssh/keys.py [hunk 1]
```diff
@@ -229,7 +302,17 @@ def _fromString_PUBLIC_OPENSSH(cls, data):
...
-        blob = base64.decodestring(data.split()[1])
+        #EC keys can come in multiple formats.
+        if data.startswith('ecdsa'):
+            newKey = cls(load_ssh_public_key(data, default_backend()))
+            newKey.ecKeyName = data[:19]
+            return newKey
+        elif 'ecdsa' in data:
+            # This string doesn't seem to be supported by any of cryptgraphy.io's utility functions.
+            # So do it the long way by grabbing the bytes we need and passing them on.
+            blob = re.search(".*?nist.\d{3}(.*)", data, re.DOTALL).group(1)
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: It strikes me that doing a regex over the entire key blob might be the source of some of your unreliability.  Surely the octets `nist` might appear at some point in the key that has nothing to do with the type.

File: twisted/conch/ssh/keys.py [hunk 2]
```diff
@@ -816,11 +931,20 @@ def type(self):
...
+        elif isinstance(
+                self._keyObject, (ec.EllipticCurvePublicKey, ec.EllipticCurvePrivateKey)):
+            return 'EC'
         else:
             raise RuntimeError(
                 'unknown type of object: %r' % (self._keyObject,))
 
 
+    def getECKeyName(self):
+            if hasattr(self, 'ecKeyName'):
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: nitpick: incorrect indentation

File: twisted/conch/ssh/keys.py [hunk 3]
```diff
@@ -816,11 +931,20 @@ def type(self):
...
             return 'DSA'
+        elif isinstance(
+                self._keyObject, (ec.EllipticCurvePublicKey, ec.EllipticCurvePrivateKey)):
+            return 'EC'
         else:
             raise RuntimeError(
                 'unknown type of object: %r' % (self._keyObject,))
 
 
+    def getECKeyName(self):
```
Comment (glyph) [inline twisted/conch/ssh/keys.py:878]: nitpick: no docstring

File: twisted/conch/ssh/keys.py [hunk 4]
```diff
@@ -919,6 +1059,10 @@ def blob(self):
             return (common.NS(b'ssh-dss') + common.MP(data['p']) +
                     common.MP(data['q']) + common.MP(data['g']) +
                     common.MP(data['y']))
+        elif type == 'EC':
+            b = (common.NS(data["curve"]) + common.NS(data["curve"][-8:]) + 
+                    common.NS(data["n"]))
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: nitpick: over-indented

File: twisted/conch/ssh/keys.py [hunk 5]
```diff
@@ -1002,6 +1154,12 @@ def _toString_OPENSSH(self, extra):
 
         @rtype: L{bytes}
         """
+        
+        # No support for EC keys yet.
```
Comment (glyph) [inline twisted/conch/ssh/keys.py:1071]: If we're going to have a comment like this in the code, it should include a link to the ticket for implementing the functionality in question, rather than just a vague apology :).

File: twisted/conch/ssh/keys.py [hunk 6]
```diff
@@ -1194,12 +1380,33 @@ def verify(self, signature, data):
...
+            r = int_from_bytes(rstr, 'big')
+            s = int_from_bytes(sstr, 'big')
+            signature = encode_dss_signature(r, s)
+
+            k = self._keyObject
+            if not self.isPublic():
+                k = k.public_key()
+
+            #Which hash to use depends on the key size.
+            if self._keyObject.curve.key_size <= 256:
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: This code is repeated from above.  I think we need a `_curve2hash` utility function or something.  Although I'm almost relieved that this is the case: when you have a big if/elif tree that all sets the same variable, it's nicer to encapsulate it as a function, because `h = foo()` only sets `h` once, whereas if you get a big tree you have to ensure that you didn't misspell `h` in any of its branches when you read it.

File: twisted/conch/ssh/keys.py [hunk 7]
```diff
@@ -1194,12 +1380,33 @@ def verify(self, signature, data):
...
+                h = hashes.SHA512()
+                
+            verifier = k.verifier(signature, ec.ECDSA(h))
         else:
             raise BadKeyError("unknown key type %s" % (self.type(),))
 
         verifier.update(data)
         try:
             verifier.verify()
+            return True
```
Comment (glyph) [inline twisted/conch/ssh/keys.py:1285]: Why this change to the behavior here?  This makes the `else` branch dead code.

File: twisted/conch/ssh/transport.py [hunk 8]
```diff
@@ -1825,7 +1988,9 @@ def decryptor(self):
 MSG_KEX_DH_GEX_INIT = 32
 MSG_KEX_DH_GEX_REPLY = 33
 
-
+#Unused but here for reference and possibly future use.
```
Comment (glyph) [inline twisted/conch/ssh/transport.py]: Please don't add unused code.  File a ticket for adding it in the future :)

File: twisted/conch/ssh/transport.py [hunk 9]
```diff
@@ -1539,6 +1700,8 @@ def _ssh_KEXDH_REPLY(self, packet):
     def ssh_KEX_DH_GEX_GROUP(self, packet):
         """
         This handles different messages which share an integer value.
+       
+        If it is ECDH, call C{_ssh_KEX_ECDH_REPLY} and pass along the packet.
```
Comment (glyph) [inline twisted/conch/ssh/transport.py]: This is an ostensibly public method, whose docstring is referencing a private implementation detail.  Why is it interesting to the caller to know this?  A maintainer could just read the code and see what it calls.  It doesn't talk about `_ssh_KEXDH_REPLY`, so this is also asymmetrical.

**Comment by the0id** [comment]
Thanks for the feed back, I'll take these changes.

As for the regex, I hear what you're saying but it's the best solution I could think of for the circumstances.

Because of the different source formats of the bytes that get passed to that function, the nist identifier isn't at a fixed position, so it needs to be searched for.  The regex expression is non-greedy so it should find the first instance of the nist-XXX identifier and then grab the rest of the data as the blob.

It's possible to do an indexOf search, but I felt that would be less reliable and more prone to false positives.

**Comment by the0id** [comment]
I was also wondering what to do about the trial-out.txt  The submission guidelines say I need to include the output of the trial script.  What's the best way to do that?

**Comment by the0id** [comment]
Ok this should be just ECDH with no ECDSA.

As far as I know everything should be working with the exception that one trial fails using python 2.7.  test_mismatchedHostKey fails on line 728 of _synchtest.py saying that it expected a failure result, but found no result instead.

I've looked into this failure and can not figure out why it happens.  All of my tests of connecting to OpenSSH servers with conch, and using Twisted SSH servers succeed every time.  So I went ahead and pushed the code.

trial twisted with Python 3 has many, many errors, but none of them seem to be related to my changes.

**Comment by the0id** [comment]
I left in the error messages about not supporting printing EC keys because I felt like it was better to throw an Exception with an explanation than to just have it crash.

I going to hold off putting in a ticket for that to be fixed just in case that gets fixed first.

File: twisted/conch/ssh/transport.py [hunk 10]
```diff
@@ -1297,6 +1371,8 @@ def ssh_KEX_DH_GEX_REQUEST_OLD(self, packet):
         # another cue to decide what kind of message the peer sent us.
```
Comment (adiroiban) [inline twisted/conch/ssh/transport.py:1370]: this comment got out of sync with the code. You might want to consider repharsing the comment or moving this code into a separate method... or refactor it somehow :)

File: twisted/conch/ssh/transport.py [hunk 11]
```diff
@@ -1539,7 +1691,7 @@ def _ssh_KEXDH_REPLY(self, packet):
     def ssh_KEX_DH_GEX_GROUP(self, packet):
         """
         This handles different messages which share an integer value.
-
+
```
Comment (glyph) [inline twisted/conch/ssh/transport.py]: Nitpick: this addition of whitespace is both unnecessary and in violation of the coding standard :)

File: twisted/conch/ssh/transport.py [hunk 12]
```diff
@@ -1593,7 +1747,6 @@ def _continueKEXDH_REPLY(self, ignored, pubKey, f, signature):
             return
         self._keySetup(sharedSecret, exchangeHash)
 
-
```
Comment (glyph) [inline twisted/conch/ssh/transport.py:1596]: similarly, this removal of whitespace is both unnecessary and in violation of the coding standard :).  Twisted's vertical whitespace requirement is different than pep-8 - 3 lines separate top-level suites, 2 lines separate method-level suites, and 1 line can be used to separate sections within a method.

File: twisted/conch/ssh/transport.py [hunk 13]
```diff
@@ -1661,7 +1814,6 @@ def _continueGEX_REPLY(self, ignored, pubKey, f, signature):
...
```
Comment (glyph) [inline twisted/conch/ssh/transport.py:1664]: ditto

File: twisted/conch/ssh/keys.py [hunk 14]
```diff
@@ -1173,9 +1265,12 @@ def verify(self, signature, data):
             signatureType, signature = b'ssh-dss', common.NS(signature)
         else:
             signatureType, signature = common.getNS(signature)
+        
         if signatureType != self.sshType():
             return False
-        if self.type() == 'RSA':
+        
+        name = self.type()
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: "name" is a tiny bit confusing as a variable name here.  If you're trying to avoid multiple calls to `self.type()`, perhaps a variable name like `myType`?

File: twisted/conch/ssh/transport.py [hunk 15]
```diff
@@ -1230,6 +1236,73 @@ def ssh_KEXINIT(self, packet):
...
+        h.update(NS(self.ourVersionString))
+        h.update(NS(self.otherKexInitPayload))
+        h.update(NS(self.ourKexInitPayload))
+        h.update(NS(pubHostKey.blob()))
+        h.update(NS(pktPub))
+
+        #Get the base curve info
+        try:
+            shortKex = re.search("(nist[kpbt]\d{3})$", self.kexAlg).group(1)
+        except:
```
Comment (glyph) [inline twisted/conch/ssh/transport.py]: You should be expecting a specific type of exception from this regular expression, not a blanket `except`.  For example, let's say a future maintainer comes through and accidentally changes `self.kexAlg` to `self.kxeAlg`; now this would raise an `AttributeError`, which would tell us `UnsupportedAlgorithm`.  In general you should catch the most specific exception type possible :).

File: twisted/conch/ssh/transport.py [hunk 16]
```diff
@@ -1470,17 +1546,40 @@ def ssh_KEXINIT(self, packet):
...
         if SSHTransportBase.ssh_KEXINIT(self, packet) is None:
             # Connection was disconnected while doing base processing.
             # Maybe no common protocols were agreed.
             return
+        # Are we using ECDH?
+        if _kex.isEllipticCurve(self.kexAlg):
+            #Find the base curve info
+            try:
+                shortKex = re.search("(nist[kpbt]\d{3})$", self.kexAlg).group(1)
+            except:
```
Comment (glyph) [inline twisted/conch/ssh/transport.py]: See the other comment about bare-word `except`

File: twisted/conch/ssh/transport.py [hunk 17]
```diff
@@ -1499,6 +1598,59 @@ def ssh_KEXINIT(self, packet):
...
+
+        h = _kex.getHashProcessor(self.kexAlg)()
+        h.update(NS(self.ourVersionString))
+        h.update(NS(self.otherVersionString))
+        h.update(NS(self.ourKexInitPayload))
+        h.update(NS(self.otherKexInitPayload))
+        h.update(NS(self.theirECHost))
+
+        #Take the provided public key and transform it into a format for the cryptography module
+        self.theirECPub = ec.EllipticCurvePublicNumbers.from_encoded_point(self.curve, pktPub).public_key(default_backend())
```
Comment (glyph) [inline twisted/conch/ssh/transport.py]: Minor, optional nitpick: it seems odd to me that we stop to load the curve mid-way through computing the hash (especially to do a side-effect).  I'd find this method easier to read if the side-effects (all the `self.x = y`) were isolated to one place - ideally at the _end_ of the method, after we have finished validating everything we're going to load.  I confess I don't fully understand the consequences of saving `theirECPub` and `theirECHost` and then raising `InvalidSignature` before calling `_keySetup`, but in general this is a recipe for bugs.

Comment (glyph) [inline twisted/conch/ssh/transport.py]: Non-optional nitpick: if you're going to add new attributes (`theirECHost`, `theirECPub`) you have to add `@ivar` descriptions to the class docstring.  However - do these attributes even need to exist?  I don't see them used anywhere outside this method.  Could they just be local variables?

File: twisted/conch/ssh/_kex.py [hunk 18]
```diff
@@ -153,6 +198,19 @@ def getKex(kexAlgorithm):
     return _kexAlgorithms[kexAlgorithm]
 
 
+def isEllipticCurve(kexAlgorithm):
```
Comment (glyph) [inline twisted/conch/ssh/_kex.py:203]: I can see that you're just copying the pattern found below, with `isFixedGroup`, but that was probably not a great pattern and the implementation could just call `_IEllipticCurveExchangeKexAlgorithm.providedBy` directly.  (Optional feedback since this doesn't touch any public interfaces, so it's purely a matter of taste.)

File: twisted/conch/ssh/factory.py [hunk 19]
```diff
@@ -58,7 +58,7 @@ def buildProtocol(self, addr):
                     'because we cannot find moduli file')
             t.supportedKeyExchanges = [
                 kexAlgorithm for kexAlgorithm in t.supportedKeyExchanges
-                if _kex.isFixedGroup(kexAlgorithm)]
+                if _kex.isFixedGroup(kexAlgorithm) or (hasattr(_kex, "isEllipticCurve") and _kex.isEllipticCurve(kexAlgorithm))]
```
Comment (glyph) [inline twisted/conch/ssh/factory.py]: Is this `hasattr` necessary?  From what I can see, `isEllipticCurve` is not conditionally defined.

File: twisted/conch/ssh/keys.py [hunk 20]
```diff
@@ -39,6 +42,47 @@
 from twisted.python.deprecate import deprecated, getDeprecationWarningString
 from twisted.python.versions import Version
 
+# Curve lookup table
+curveTable = {
```
Comment (glyph) [inline twisted/conch/ssh/keys.py:47]: `curveTable` and `oidTable` are new public APIs according to Twisted's compatibility policy.  Do we actually need to expose these to applications?  I suspect not.

File: twisted/conch/ssh/keys.py [hunk 21]
```diff
@@ -39,6 +42,47 @@
...
+    '1.3.132.0.26' : ec.SECT233K1(),
+    '1.3.132.0.27' : ec.SECT233R1(),
+    '1.3.132.0.16' : ec.SECT283K1(),
+    '1.3.132.0.36' : ec.SECT409K1(),
+    '1.3.132.0.37' : ec.SECT409R1(),
+    '1.3.132.0.38' : ec.SECT571K1()
+    }
+
+#Return the curve table dictionary for nist name lookup.
+def getCurveTable():
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: Given that `curveTable` and `oidTable` are both public on their own, these accessors serve no purpose and should just be deleted.

File: twisted/conch/ssh/keys.py (hunk 20 above)
Comment (glyph) [inline twisted/conch/ssh/keys.py:47]: If they are going to remain (whether public or private) each should have a `@var` declaration in the module docstring.

File: twisted/conch/ssh/keys.py [hunk 22]
```diff
@@ -639,27 +696,33 @@ def __repr__(self):
...
-                    o = o[:-1]
-                lines.append('\t' + o)
-        lines[-1] = lines[-1] + '>'
-        return '\n'.join(lines)
+        if self.type() == 'EC':
+            if self.isPublic():
+                return self._keyObject.public_bytes(serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo)
+            else:
+                return self._keyObject.private_bytes(serialization.Encoding.PEM, serialization.PrivateFormat.TraditionalOpenSSL, serialization.NoEncryption())
+        else:
```
Comment (glyph) [inline twisted/conch/ssh/keys.py:695]: Optional nitpick - this is more of a trick you can use in the future - given that all the branches under `self.type() == 'EC'` end up exiting the function, you can reduce the number of diff lines quite a bit by not adding an `else` and not indenting the rest of the implementation.  This means that the reviewer will just see the added lines instead of having to try to see if any of the rest of the implementation has changed.

File: twisted/conch/ssh/keys.py [hunk 23]
```diff
@@ -1059,6 +1140,12 @@ def _toString_LSH(self):
         """
         data = self.data()
         type = self.type()
+
+        # No support for EC keys yet.
+        if self.type() == 'EC':
+            raise UnsupportedAlgorithm("toString() does not support  Elliptic Curves yet.")
+            return
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: This `return` is dead code.

File: twisted/conch/ssh/keys.py [hunk 24]
```diff
@@ -1118,6 +1205,12 @@ def _toString_AGENTV3(self):
             elif self.type() == 'DSA':
                 values = (data['p'], data['q'], data['g'], data['y'],
                           data['x'])
+            # I assume this is what's needed
+            # But commented out because it hasn't been tested yet.
+            elif self.type() == 'EC':
+                raise UnsupportedAlgorithm("toString() does not support Elliptic Curves yet.")
+                return
```
Comment (glyph) [inline twisted/conch/ssh/keys.py]: Similarly: dead code.

File: twisted/conch/ssh/transport.py (hunk 17 above)
Comment (the0id) [inline twisted/conch/ssh/transport.py]: The breaking up of updating the hash is a left over of my experimentation, but I agree it should be changed.

Raising InvalidSignature was to make troubleshooting a little easier by identifying what went wrong and where as close to the exception as possible.  This was a big headache in implementing ECDH so I thought I'd be nice and help the next poor sap who had the same trouble I did. :) 

theirECHost has to exist, but not as self.theirECHost.  I was following the convention I saw else where of defining parts of the key were saved as self variables (e.g. self.x and self.g from ssh_KEXINIT).

Everything I mention in the comment isn't necessarily required so I'm happy to change what you see fit.

File: twisted/conch/ssh/_kex.py (hunk 18 above)
Comment (the0id) [inline twisted/conch/ssh/_kex.py:203]: If it's a matter of taste then I'll keep it as is.  It seems to fit the convention used by isFixedGroup and in my eyes easier to understand.

File: twisted/conch/ssh/keys.py (hunk 20 above)
Comment (the0id) [inline twisted/conch/ssh/keys.py]: My programming teacher told me to NEVER access variables outside of a class directly.  He was very strict about this, and and for good reason.

But this is python, and normal rules no longer apply. :)

File: twisted/conch/ssh/keys.py (hunk 21 above)
Comment (the0id) [inline twisted/conch/ssh/keys.py]: My programming teacher told me to NEVER directly access variables outside of a class.  He was very strict about this, and for good reason.

But this is python, and normal rules no longer apply. :)

File: twisted/conch/ssh/factory.py (hunk 19 above)
Comment (the0id) [inline twisted/conch/ssh/factory.py]: At one point it was otherwise an exception was thrown.

Whatever was causing it seems to have fixed itself.

File: twisted/conch/ssh/keys.py (hunk 20 above)
Comment (the0id) [inline twisted/conch/ssh/keys.py:47]: I added some @ivar declarations at the top of the module, but I don't know if it was done correctly.

File: twisted/conch/ssh/keys.py [hunk 25]
```diff
@@ -1118,6 +1205,12 @@ def _toString_AGENTV3(self):
             elif self.type() == 'DSA':
                 values = (data['p'], data['q'], data['g'], data['y'],
                           data['x'])
+            # I assume this is what's needed
+            # But commented out because it hasn't been tested yet.
```
Comment (glyph) [inline twisted/conch/ssh/keys.py:1197]: I think this comment is out of sync; but generally, don't comment stuff out; http://nedbatchelder.com/text/deleting-code.html :)

**Comment by glyph** [comment]
Thanks for keeping pace with my changes as I was suggesting them - I had to do the review slowly over the course of the day and it's nice to see that it's mostly addressed already :)

**Comment by glyph** [comment]
> I was also wondering what to do about the trial-out.txt The submission guidelines say I need to include the output of the trial script. What's the best way to do that?

Where did you see this in the contribution guidelines?  This must be some text left over from before we had continuous integration and could just look at the test output on our own buildbot :).  In any case it is outdated and you can ignore this requirement - just make sure that the github build statuses are passing.

**Comment by glyph** [comment]
> trial twisted with Python 3 has many, many errors, but none of them seem to be related to my changes.

This is why there's an `admin/run-python3-tests` script.  Not all the tests are expected to pass on python3, and that script will honor the expected set.

**Comment by glyph** [comment]
Finally - please resolve the conflicts in the branch.

File: twisted/conch/ssh/keys.py (hunk 25 above)
Comment (the0id) [inline twisted/conch/ssh/keys.py:1197]: That's fair.  I had that in there to share a possible solution, but it if could complicate or confuse things I have no problem taking it out.

**Comment by the0id** [comment]
I tried to find where I saw that the trial output must be included, but couldn't.

So either I found old information, or I just made it up. :)

**Comment by adiroiban** [comment]
> So either I found old information, or I just made it up. :)

It might be some very, very old information.

Now we have travis-ci and appveyour so we can see the results of trial as running on your branch.

**Comment by the0id** [comment]
I can't see what conflicts exist in my commits.  I can see the travis-ci and appveyor checks box on the page, but the only information I can get is that one check failed and one check is waiting for status to be reported.

I tried clicking all over it and all I can do is hide the checks.  Am I missing something?

So I pulled the latest Twisted trunk and merged in those changes into my file, which hopefully should resolve any conflicts.

**Comment by glyph** [comment]
Hrm. I wonder if this is not running CI because I created the PR but I don't own the remote? I am going to try closing and reopening to see if that jogs it.

**Comment by glyph** [comment]
Aaaaand now it has conflicts that need merging.

**Comment by the0id** [comment]
I'm not noticing any change on my end.  It is still showing the same box as before saying that one check has failed and the other is waiting for status to be reported.

Would it be an ownership issue? Should I just open a new PR?

**Comment by glyph** [comment]
@the0id - yeah, let's just close this one out and open a new one.  I just pushed a branch myself and it still isn't triggering.

**Comment by glyph** [comment]
@the0id can you link to the new one here when it's open?

**Comment by the0id** [comment]
Sure: https://github.com/twisted/twisted/pull/422
